component/module/git/git.go

```diff
@@ -0,0 +1,302 @@
+	Revision   string `river:"revision,attr"`
+	Path       string `river:"path,attr"`
+
+	PullFrequency time.Duration `river:"pull_frequency,attr,optional"`
+
+	Arguments map[string]any `river:"arguments,attr,optional"`
+}
+
+// DefaultArguments holds default settings for Arguments.
+var DefaultArguments = Arguments{
+	PullFrequency: time.Minute,
+}
+
+var _ river.Unmarshaler = (*Arguments)(nil)
+
+// UnmarshalRiver implements river.Unmarshaler and applies default settings to
+// args.
+func (args *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*args = DefaultArguments
+
+	type arguments Arguments
+	return f((*arguments)(args))
+}
+
+// Exports are emitted by the module.git component.
+type Exports struct {
+	Exports map[string]any `river:"exports,attr"`
+}
+
+// Component implements the module.git component.
+type Component struct {
+	opts component.Options
+	log  log.Logger
+	ctrl *flow.Flow
+
+	mut      sync.RWMutex
+	repo     *vcs.GitRepo
+	repoOpts vcs.GitRepoOptions
+	args     Arguments
+
+	argsChanged chan struct{}
+
+	healthMut sync.RWMutex
+	health    component.Health
+}
+
+var (
+	_ component.Component       = (*Component)(nil)
+	_ component.HealthComponent = (*Component)(nil)
+	_ component.HTTPComponent   = (*Component)(nil)
+)
+
+// New creates a new module.git component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	// TODO(rfratto): replace these with a tracer/registry which properly
+	// propagates data back to the parent.
+	flowTracer, _ := tracing.New(tracing.DefaultOptions)
+	flowRegistry := prometheus.NewRegistry()
+
+	c := &Component{
+		opts: o,
+		log:  o.Logger,
+
+		ctrl: flow.New(flow.Options{
+			ControllerID: o.ID,
+			LogSink:      logging.LoggerSink(o.Logger),
+			Tracer:       flowTracer,
+			Reg:          flowRegistry,
+
+			DataPath:       o.DataPath,
+			HTTPPathPrefix: o.HTTPPath,
+			HTTPListenAddr: o.HTTPListenAddr,
+
+			OnExportsChange: func(exports map[string]any) {
+				o.OnStateChange(Exports{Exports: exports})
+			},
+		}),
+
+		argsChanged: make(chan struct{}, 1),
+	}
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	go c.ctrl.Run(ctx)
+
+	var (
+		ticker  *time.Ticker
+		tickerC <-chan time.Time
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-c.argsChanged:
+			c.mut.Lock()
+			{
+				level.Info(c.log).Log("msg", "updating repository pull frequency", "new_frequency", c.args.PullFrequency)
+
+				if c.args.PullFrequency > 0 {
+					if ticker == nil {
+						ticker = time.NewTicker(c.args.PullFrequency)
+						tickerC = ticker.C
+					} else {
+						ticker.Reset(c.args.PullFrequency)
+					}
+				} else {
+					if ticker != nil {
+						ticker.Stop()
+					}
+					ticker = nil
+					tickerC = nil
+				}
+
+			}
+			c.mut.Unlock()
```

Comment:
This code is pretty gross. I'd prefer cleaning this up if we want to move forward with the module loader.